### ISSUE TITLE
Uploads sometimes get stuck

### ISSUE THREAD

**Issue by hamann** [epiccastle/spire#38]
Uploads sometimes get stuck. 

```txt
src/datadog.clj:24 (apt-repo :present {:filename "datadog", :repo "deb https://apt.datadoghq.com/ stable 7"}) vagrant@localhost:54321
src/datadog.clj:26 (shell {:cmd "apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 A2923DFF56EDA6E76E55E492D3A80E30382E94DE"}) vagrant@localhost:54321
src/datadog.clj:27 (apt :update) vagrant@localhost:54321
src/datadog.clj:28 (apt :install "datadog-agent") vagrant@localhost:54321
src/datadog.clj:31 (upload {:content (selmer "uploads/datadog/datadog.yaml" {:api_key (api-key)}), :dest "/etc/datadog-agent/datadog.yaml"}) vagrant@localhost:54321
src/datadog.clj:34 (shell {:cmd "gpasswd -a dd-agent systemd-journal"}) vagrant@localhost:54321
src/datadog.clj:35 (shell {:cmd "mkdir -p /etc/datadog-agent/conf.d/journal.d"}) vagrant@localhost:54321
src/datadog.clj:36 (upload {:content (selmer "uploads/datadog/journal.d/conf.yaml" {}), :dest "/etc/datadog-agent/conf.d/journal.d/conf.yaml"}) vagrant@localhost:54321
src/datadog.clj:39 (upload {:content (selmer "uploads/datadog/conf.d/nagios_runner.yaml" {}), :dest "/etc/datadog-agent/conf.d/nagios_runner.yaml"}) vagrant@localhost:54321
src/datadog.clj:42 (upload {:content (selmer "uploads/datadog/checks.d/nagios_runner.py" {}), :dest "/etc/datadog-agent/checks.d/nagios_runner.py"}) vagrant@localhost:54321
src/datadog.clj:45 (shell {:cmd "mkdir -p /etc/datadog-agent/scripts"}) vagrant@localhost:54321
src/datadog.clj:46 (upload {:content (selmer "uploads/datadog/scripts/check_ssacli" {}), :dest "/etc/datadog-agent/scripts/check_ssacli"})
vagrant@localhost:54321 [String Data]|=============================================================================================================================================| 100%
```

The file was uploaded, but spire runs forever

```sh
root@ubuntu-bionic:~$ stat /etc/datadog-agent/scripts/check_ssacli
  File: /etc/datadog-agent/scripts/check_ssacli
  Size: 6158            Blocks: 16         IO Block: 4096   regular file
Device: 801h/2049d      Inode: 279833      Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2020-04-24 10:26:44.613938116 +0000
Modify: 2020-04-24 10:27:23.381938116 +0000
Change: 2020-04-24 10:27:23.381938116 +0000
 Birth: -

root@ubuntu-bionic:~$ ps aux
...
vagrant  23423  0.0  0.0  13312  3072 ?        Ss   10:27   0:00 bash -c umask 0000;sudo -S -p '' -u 'root'  scp   -t /etc/datadog-agent/scripts/check_ssacli
root     23424  0.0  0.1  63968  4272 ?        S    10:27   0:00 sudo -S -p  -u root scp -t /etc/datadog-agent/scripts/check_ssacli
root     23425  0.0  0.0  14924  2252 ?        S    10:27   0:00 scp -t /etc/datadog-agent/scripts/check_ssacli

...
```

**Comment by retrogradeorbit** [comment]
Is this intermittent? does it always happen in the same file? the file is complete and correctly uploaded?

I will try and recreate it with vagrant, sudo and upload.

**Comment by hamann** [comment]
It’s always the same file and occurs on two different hosts, so I think and hope it’s reproducable. After upload the file is complete from a content view. Should I share it?

**Comment by retrogradeorbit** [comment]
if you are able to share it that would be great. What are the two hosts? vagrant and a real machine?

**Comment by retrogradeorbit** [comment]
I might be able to recreate it. Using a 6158 byte text file, to ubuntu bionic, under sudo, I get a scp protocol error. An exception, not a hang. But may indicate a problem. Investigating further...

**Comment by hamann** [comment]
The local file and the uploaded file have different md5sums, so I made a `diff`

```txt
190c190
< $np->nagios_exit( $np->check_messages() );
\ No newline at end of file
---
> $np->nagios_exit( $np->check_messages() );
```

[check_ssacli.txt](https://github.com/epiccastle/spire/files/4531558/check_ssacli.txt)

**Comment by hamann** [comment]
> if you are able to share it that would be great. What are the two hosts? vagrant and a real machine?

Yes, dev and prod

**Comment by hamann** [comment]
The upload finishes with `:src "file"` instead of `:content (selmer ..)`

Btw, `:src` and `:content` treat paths differently. `:src` expects the file relative to root, whereas `:content` expects it relative to script.

**Comment by retrogradeorbit** [comment]
I have found that the scp upload code is misreporting the size of the pending transfer to scp when it is inisde `:content`. Works as `:src` because it sends the filesize. Whereas with `:content` it sends the string length. This isn't the size of the binary representation of the file, so for example a file with non ascii characters or multibyte unicode will have a different string length to binary size. This explains why my testing (using plain ascii works)

In addition, :src should be relative to the script like `selmer` is. I will fix that aswell.

**Comment by retrogradeorbit** [comment]
should be fixed in 0.1.0-alpha.6

@hamann please reopen if it is still hanging!
